### PR TITLE
Bucket change during integration  

### LINKED ISSUE #1983: Add the ability to handle accumulations and max/min computations over multiple time windows
## Description
 
The RRFS plans to generate output at 15 minute intervals for an initial portion of its forecast, and then generate hourly output for the remainder of the forecast.  We'd like to have the ability to simultaneously track 15 minute (or even shorter) and 60 minute periods to produce the most relevant output at specific model output times.
 

## Solution
 
Ideally the fhzero and avg_max_length namelist variables could be prescribed as a list of zeroing out or averaging lengths of time rather than single values.

## Alternatives
 
We've considered the possibility of restarting the model when switching the output frequency from 15 minute to hourly (with a change to the fhzero and avg_max_length namelist variables, but this comes with a performance overhead.

### PR DESCRIPTION
**Pull request by junwang-noaa**
## Description

(Instructions: this, and all subsequent sections of text should be removed and filled in as appropriate.)
Provide a detailed description of what this PR does.  
What bug does it fix, or what feature does it add?  
Is a change of answers expected from this PR?  

RRFS requested to allow bucket change during integration. This is initial code updates for that capability. 

At this time, the bucket (fhzero) is updated the forecast side, we need to update attribute at forecast time to pass the change to write grid component.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues should always be created before starting work on a PR branch!)
- fixes #<issue_number>
- fixes noaa-emc/fv3atm/issues/<issue_number>
- fixes https://github.com/ufs-community/ufs-weather-model/issues/1983

## Testing

How were these changes tested?  
What compilers / HPCs was it tested with?  
Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)  
Have the ufs-weather-model regression test been run? On what platform?  
- Will the code updates change regression test baseline? If yes, why? Please show the baseline directory below.
- Please commit the regression test log files in your ufs-weather-model branch

test with regional_control test. The code is backward compatible, there is no impact on current tests.

## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
Those branches should have matching names (ideally)

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs
- waiting on noaa-emc/nems/pull/<pr_number>
- waiting on noaa-emc/fv3atm/pull/<pr_number>

### CODE REVIEW

**Comment by MatthewPyle-NOAA** [comment]
I think one item for RRFS got lost along the way - the need for tracking multiple bucket lengths at the same time (namely 15 minute and 1 hour).  But we've recently decided to avoid generating most "bucket" type fields from the RRFS 15 minute output, and just use an external code to generate 15 minute QPF buckets from the running QPF total.   Sorry for the confusion.

**Comment by junwang-noaa** [comment]
@MatthewPyle-NOAA Thanks for clarification. Currently model does not support multiple buckets at the same time. In this branch, model only supports a single bucket, but the bucket length could change. I tested the branch with non-integer bucket length (0.25 hour), this history has correct bucket length, but UPP does not support non-integer bucket.

Here is the output from a test run:

history file:
```
// global attributes:
                :cen_lat = 38.f ;
                :cen_lon = -120.f ;
                :dtp = 36.f ;
                :dx = 3000.f ;
                :dy = 3000.f ;
                :fhzero = 0.25f ;
                :grid = "lambert_conformal" ;
                :grid_id = 1LL ;
```
The post grib file:
```
851:33388831:vt=2022082403:surface:3-3 hour acc fcst:APCP Total Precipitation [kg/m^2]:
    ndata=61440:undef=511:mean=0:min=0:max=0
    grid_template=30:winds(N/S):
        Lambert Conformal: (320 x 192) input WE:SN output WE:SN res 0
        Lat1 35.299999 Lon1 234.750000 LoV 240.000000
        LatD 38.000000 Latin1 38.000000 Latin2 38.000000
        LatSP 0.000000 LonSP 0.000000
        North Pole (320 x 192) Dx 3000.000000 m Dy 3000.000000 m mode 0
```

**Comment by MatthewPyle-NOAA** [comment]
Thanks for your work on getting sub-hourly buckets working (at least in the model).  I'm guessing we'll use this functionality in some fashion with RRFS for our 15 minute output.

File: atmos_model.F90 [hunk 1]
```diff
@@ -533,12 +533,12 @@ subroutine atmos_model_init (Atmos, Time_init, Time, Time_step)
...
 !--- local variables ---
   integer :: unit, i
-  integer :: mlon, mlat, nlon, nlat, nlev, sec
+  integer :: mlon, mlat, nlon, nlat, nlev, sec, sec_lastfhzerofh
   integer :: ierr, io, logunit
   integer :: tile_num
   integer :: isc, iec, jsc, jec
   real(kind=GFS_kind_phys) :: dt_phys
-  logical              :: p_hydro, hydro
+  logical              :: p_hydro, hydro, tmpflag
```
Comment (BrianCurtis-NOAA) [inline atmos_model.F90]: if this variable is going to stay in the code, you may want to choose a more descriptive name?

File: io/post_fv3.F90 [hunk 2]
```diff
@@ -628,15 +627,15 @@ subroutine set_postvars_fv3(wrt_int_state,grid_id,mype,mpicomp)
...
       tclod   = tprec
       trdlw   = tprec
       trdsw   = tprec
       tsrfc   = tprec
       tmaxmin = tprec
       td3d    = tprec
-!      if(mype==0)print*,'MP_PHYSICS= ',imp_physics,'tprec=',tprec,'tclod=',tclod, &
-!       'dtp=',dtp,'tmaxmin=',tmaxmin,'jsta=',jsta,jend,im,jm
+      if(mype==0)print*,'MP_PHYSICS= ',imp_physics,'tprec=',tprec,'tclod=',tclod, &
+       'dtp=',dtp,'tmaxmin=',tmaxmin,'jsta=',jsta,jend,im,jm
```
Comment (BrianCurtis-NOAA) [inline io/post_fv3.F90]: I am guessing this is probably a debugging print statement that you don't want to leave uncommented?

File: io/module_wrt_grid_comp.F90 [hunk 3]
```diff
@@ -1896,6 +1776,23 @@ subroutine wrt_run(wrt_comp, imp_state_write, exp_state_write,clock,rc)
                            fieldbundle=file_bundle, rc=rc)
         if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
 
+#if 1
```
Comment (BrianCurtis-NOAA) [inline io/module_wrt_grid_comp.F90]: this #if 1 seems like some testing section? Do you need this here?

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]

File: atmos_model.F90 (hunk 1 above)
Comment (junwang-noaa) [reply atmos_model.F90]: Updated

File: io/post_fv3.F90 (hunk 2 above)
Comment (junwang-noaa) [reply io/post_fv3.F90]: Commented out.

File: io/module_wrt_grid_comp.F90 (hunk 3 above)
Comment (junwang-noaa) [reply io/module_wrt_grid_comp.F90]: Removed

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by DusanJovic-NOAA** [comment]
@junwang-noaa @MatthewPyle-NOAA Should this change go to develop or RRFSv1 branch or maybe both?

**Comment by MatthewPyle-NOAA** [comment]
@DusanJovic-NOAA We won't be utilizing it in RRFSv1, so just develop is fine.

**Comment by DusanJovic-NOAA** [comment]
This was replaced by #835. Closing.

**Pull request closed without merging**